### ISSUE TITLE
write_install includes duplicate packages

### ISSUE THREAD

**Issue by jdblischak** [karthik/holepunch#43]
## What I observed

When multiple files import the same package, the file `install.R` created by `write_install()` runs `install.packages()` multiple times.

## What I expected

I expected that each package used in a project would be installed once in `install.R`.

## Reproducible example

``` r
packageVersion("holepunch")
#> [1] '0.1.26.9000'
packageVersion("renv")
#> [1] '0.9.2'

d <- tempfile()
dir.create(d)
f1 <- file.path(d, "f1.R")
writeLines(c("library(rmarkdown)", "library(knitr)"), con = f1)
f2 <- file.path(d, "f2.R")
writeLines("library(rmarkdown)", con = f2)
renv::dependencies(d)
#> Finding R package dependencies ... Done!
#>                                  Source   Package Require Version   Dev
#> 1 /tmp/RtmpaLehvK/file266e481c0b2e/f1.R     knitr                 FALSE
#> 2 /tmp/RtmpaLehvK/file266e481c0b2e/f1.R rmarkdown                 FALSE
#> 3 /tmp/RtmpaLehvK/file266e481c0b2e/f2.R rmarkdown                 FALSE
holepunch::get_dependencies(d)
#> Finding R package dependencies ... Done!
#> [1] "knitr"     "rmarkdown" "rmarkdown"
holepunch::write_install(d)
#> Finding R package dependencies ... Done!
#> ✔ Writing '/tmp/RtmpaLehvK/file266e481c0b2e/.binder/install.R'
install <- file.path(d, ".binder", "install.R")
readLines(install)
#> [1] "install.packages('knitr')"     "install.packages('rmarkdown')"
#> [3] "install.packages('rmarkdown')"
```

<sup>Created on 2020-03-01 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

<details>

<summary>Session info</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 3.6.3 (2020-02-29)
#>  os       Ubuntu 18.04.4 LTS          
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/New_York            
#>  date     2020-03-01                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version     date       lib source                            
#>  assertthat    0.2.1       2019-03-21 [3] CRAN (R 3.5.3)                    
#>  backports     1.1.5       2019-10-02 [3] CRAN (R 3.6.1)                    
#>  cli           2.0.2       2020-02-28 [3] CRAN (R 3.6.3)                    
#>  cliapp        0.1.0       2018-12-16 [1] CRAN (R 3.6.1)                    
#>  clisymbols    1.2.0       2017-05-21 [3] CRAN (R 3.5.1)                    
#>  crayon        1.3.4       2017-09-16 [3] CRAN (R 3.5.0)                    
#>  desc          1.2.0       2018-05-01 [3] CRAN (R 3.5.0)                    
#>  digest        0.6.25      2020-02-23 [3] CRAN (R 3.6.3)                    
#>  evaluate      0.14        2019-05-28 [3] CRAN (R 3.6.0)                    
#>  fansi         0.4.1       2020-01-08 [3] CRAN (R 3.6.2)                    
#>  fs            1.3.1.9000  2020-01-08 [1] Github (r-lib/fs@2fb543e)         
#>  gh            1.1.0       2020-01-24 [3] CRAN (R 3.6.2)                    
#>  glue          1.3.1       2019-03-12 [3] CRAN (R 3.5.3)                    
#>  highr         0.8         2019-03-20 [3] CRAN (R 3.5.3)                    
#>  hms           0.5.3       2020-01-08 [3] CRAN (R 3.6.2)                    
#>  holepunch     0.1.26.9000 2020-02-28 [1] Github (karthik/holepunch@0eec31c)
#>  htmltools     0.4.0       2019-10-04 [3] CRAN (R 3.6.1)                    
#>  httr          1.4.1       2019-08-05 [3] CRAN (R 3.6.1)                    
#>  jsonlite      1.6.1       2020-02-02 [3] CRAN (R 3.6.2)                    
#>  knitr         1.28        2020-02-06 [3] CRAN (R 3.6.2)                    
#>  lubridate     1.7.4       2018-04-11 [3] CRAN (R 3.5.0)                    
#>  magrittr      1.5         2014-11-22 [3] CRAN (R 3.5.0)                    
#>  pkgconfig     2.0.3       2019-09-22 [3] CRAN (R 3.6.1)                    
#>  prettyunits   1.1.1       2020-01-24 [3] CRAN (R 3.6.2)                    
#>  progress      1.2.2       2019-05-16 [3] CRAN (R 3.6.0)                    
#>  R6            2.4.1       2019-11-12 [3] CRAN (R 3.6.1)                    
#>  Rcpp          1.0.3       2019-11-08 [3] CRAN (R 3.6.1)                    
#>  renv          0.9.2       2019-12-09 [1] CRAN (R 3.6.2)                    
#>  reprex        0.3.0.9000  2019-07-19 [1] local                             
#>  rlang         0.4.4       2020-01-28 [3] CRAN (R 3.6.2)                    
#>  rmarkdown     2.1         2020-01-20 [1] CRAN (R 3.6.2)                    
#>  rprojroot     1.2         2017-01-16 [3] CRAN (R 3.5.0)                    
#>  selectr       0.4-2       2019-11-20 [3] CRAN (R 3.6.1)                    
#>  sessioninfo   1.1.1       2018-11-05 [3] CRAN (R 3.5.1)                    
#>  stringi       1.4.6       2020-02-17 [3] CRAN (R 3.6.2)                    
#>  stringr       1.4.0       2019-02-10 [3] CRAN (R 3.5.2)                    
#>  usethis       1.5.1.9000  2020-02-04 [1] Github (r-lib/usethis@ff34e40)    
#>  vctrs         0.2.3       2020-02-20 [3] CRAN (R 3.6.2)                    
#>  withr         2.1.2       2018-03-15 [3] CRAN (R 3.5.0)                    
#>  xfun          0.11        2019-11-12 [3] CRAN (R 3.6.1)                    
#>  xml2          1.2.2       2019-08-09 [3] CRAN (R 3.6.1)                    
#>  yaml          2.2.1       2020-02-01 [3] CRAN (R 3.6.2)                    
#> 
#> [1] /home/jdb-work/R/x86_64-pc-linux-gnu-library/3.6
#> [2] /usr/local/lib/R/site-library
#> [3] /usr/lib/R/site-library
#> [4] /usr/lib/R/library
```

</details>

## Proposal

I propose solving this by running `unique()` on the packages returned by `renv::dependencies()`. This could either be done in `get_dependencies()` or `write_install()`.

Also, is it a requirement for binder/repo2docker that each package is installed on a separate line in [install.r](https://repo2docker.readthedocs.io/en/latest/config_files.html#install-r-install-an-r-rstudio-environment)? The installation would be faster if all the packages were passed to `install.packages()` at once, so that it can figure out the dependency order.

For example, the lines below install rstudioapi twice (since it is a dependency of getPass):

```
install.packages("getPass")
install.packages("rstudioapi")
```

whereas the following only installs it once:

```
install.packages(c(
  "getPass",
  "rstudioapi"
))
```

The only added difficulty in the second approach is removing the comma from the final entry.

Are you interested in either of these proposed changes? At the upcoming [Chicago R Collaborative](https://chircollab.github.io/), I am going to be mentoring new contributors to open source R packages, and I think these changes would be ideal for a new contributor.

**Comment by annakrystalli** [comment]
Was just about to open a similar issue. `holepunch::write_install()` also results in duplicate `install.packages()` entries for the same package at times. (similar to #24). I feel the error is emanating from `holepunch::get_dependencies()`.

**Comment by karthik** [comment]
Thanks both, I'll look into it now.

**Comment by karthik** [comment]
Reading your whole comment @jdblischak 

> Are you interested in either of these proposed changes? At the upcoming Chicago R Collaborative, I am going to be mentoring new contributors to open source R packages, and I think these changes would be ideal for a new contributor.

Yes please have a new contributor submit these changes. Will do a full code review, accept and add them as a contributor.

**Comment by jdblischak** [comment]
@karthik Heads up that the [Chicago R Collaborative](https://chircollab.github.io/) was changed to a virtual event taking place this Saturday and Sunday (2020-04-18/9). Thus I hope to have a PR sent your way sometime over the next few days.

**Comment by karthik** [comment]
Thank you! Note: I ran into a small issue with the recent usethis update. Will fix on Monday. If you run into trouble over there weekend just revert to an older version of usethis and it will pass.

**Comment by jdblischak** [comment]
Closed by @denironyx in PR #48
